### PR TITLE
[flaky-test] Add log view flaky test

### PR DESCRIPTION
**Pull request by janeyx99**
Adds a log viewer for flaky test history view!
image
https://torchci-pqabz1gie-torchci.vercel.app/flakytest?name=test_forward_overlap

### CODE REVIEW

File: torchci/pages/flakytest.tsx [hunk 1]
```diff
@@ -50,6 +52,7 @@ export default function Page() {
                   })}
                 </ul>
               </div>
+              {samples?.length > 0 && <div><p>Example logs: </p> <LogViewer job={samples[0]} /></div>}
```
Comment (janeyx99) [inline torchci/pages/flakytest.tsx:55]: This just gets the first sample--maybe in the future we can put one for every job

**Review by suo — APPROVED** [review]

File: torchci/lib/fetchFailureSamples.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import getRocksetClient from "lib/rockset";
+import rocksetVersions from "rockset/prodVersions.json";
+
+import { JobData } from "./types"
+
+export default async function fetchFailureSamples(capture: string | string[]): Promise<JobData[]> {
```
Comment (zengk95) [inline torchci/lib/fetchFailureSamples.ts:6]: I feel like we should make it either a string or string array, not both.

File: torchci/pages/api/failure/[capture].ts [hunk 3]
```diff
@@ -1,41 +1,26 @@
...
-  );
+  const samples = await fetchFailureSamples(req.query.capture);
 
   const jobCount: {
     [jobName: string]: number;
   } = {};
 
-  for (const result of samples.results!) {
-    jobCount[result.name] = (jobCount[result.name] || 0) + 1;
+  for (const result of samples!) {
```
Comment (zengk95) [inline torchci/pages/api/failure/[capture].ts:16]: i don't think you need an exclaimation mark here or in the other places.

File: torchci/pages/api/flaky-tests/flakytest.ts [hunk 4]
```diff
@@ -18,5 +23,24 @@ export default async function handler(
...
+  const flakySamples: {
+    [capture: string]: JobData[];
+  } = {};
+
+  // only get log view for fewer than 5 flaky tests (to not spam query calls), which should be every case as we
+  // now limit on test name
+  console.debug(`Retrieved ${flakyTests.length} flaky tests`);
+  if (flakyTests.length < 5) {
+    const unfulfilledPromises = flakyTests.map(async function(flakyTest) {
+      return await fetchFailureSamples(getFlakyTestCapture(flakyTest));
```
Comment (zengk95) [inline torchci/pages/api/flaky-tests/flakytest.ts:36]: if we're awaiting them in here, then they aren't really unfufilled; they actually get executed. I think you can get rid of the async part on line 35 and remove the await on here.

File: torchci/pages/flakytest.tsx [hunk 5]
```diff
@@ -29,8 +30,9 @@ export default function Page() {
         <div>Loading...</div>
       ) : (
         (data.flakyTests as FlakyTestData[]).map((test) => {
+          const samples = data.flakySamples[`${test.name}, ${test.suite}`];
```
Comment (zengk95) [inline torchci/pages/flakytest.tsx:33]: can we import this method from flakytest.ts. Not sure if everyone will know how to encode these keys.

File: torchci/lib/fetchFailureSamples.ts (hunk 2 above)
Comment (janeyx99) [reply torchci/lib/fetchFailureSamples.ts:6]: I'm gonna leave it cuz this was from the code before--maybe can fix in a different PR

**Pull request merged**